evaluation.py

```diff
@@ -82,14 +82,20 @@ def get_metrics(x, ranks):
     topic_nodes = nodes[nodes["kind"] == "topic"].reset_index(drop=True)
     content_nodes = nodes[nodes["kind"] != "topic"].reset_index(drop=True)
     
-    topic_nodes_embeddings = embeddings.loc[topic_nodes["node_id"], :].reset_index(drop=True)
-    content_nodes_embeddings = embeddings.loc[content_nodes["node_id"], :].reset_index(drop=True)
+    topic_nodes_embeddings = embeddings.loc[topic_nodes["node_id"]].iloc[:,:-1]
+    topic_nodes_weights = embeddings.loc[topic_nodes["node_id"]].iloc[:,-1]
+    content_nodes_embeddings = embeddings.loc[content_nodes["node_id"]].iloc[:,:-1]
+
     print("done")
     sys.stdout.flush()
 
     print("Calculating cosine similarity...", end=" ")
     sys.stdout.flush()
     cos_sim = cosine_similarity(topic_nodes_embeddings.to_numpy(), content_nodes_embeddings.to_numpy())
+    cos_sim = pd.DataFrame(cos_sim, index=topic_nodes_embeddings.index)
+    cos_sim = cos_sim.multiply(topic_nodes_weights, axis="index")
+    cos_sim = cos_sim.groupby(cos_sim.index).max()
```

Comment:
Oh, very interesting! I had been thinking of this differently, but this may be better.

I had been imagining each content node's score as the _weighted average_ of its cosine similarities with the topic's embeddings, instead of the _maximum weighted_ cosine similarity (if I'm interpreting this correctly).

Your approach (max) likely better supports diversity (as it only requires that a target content item be near *one* of the topic embeddings to score highly), whereas my approach (average) would place stronger emphasis on content items that sit somewhere in between multiple topic embeddings.

I guess we'll have to see in practice, empirically, which of those approaches gives better ranking results. Maybe we could have the evaluation framework output separate performance metrics for each of the following methods?
- Max of the weighted cosine similarities (like you have here)
- Weighted sum of the cosine similarities (as I propose above)
- Cosine similarities with a weighted sum of the topic vectors (basically, reverting back to the single topic embedding approach by computing a weighted centroid of the topic embeddings)

Reply:
That's interesting! I'll add the other two approaches.